src/compiler/sys/in-memory-fs.ts

```diff
@@ -1,77 +1,226 @@
+   * not the entire path. Basically for "node_modules".
+   */
+  excludeDirNames?: string[];
+  /**
+   * Extensions we know we can avoid. Each extension
+   * should include the `.` so that we can test for both
+   * `.d.ts.` and `.ts`. If `excludeExtensions` isn't provided it
+   * doesn't try to exclude anything. This only checks against
+   * the filename, not directory names when recursive.
+   */
+  excludeExtensions?: string[];
+}
+
+/**
+ * A result from a directory read operation
+ */
+interface FsReaddirItem {
+  absPath: string;
+  relPath: string;
+  isDirectory: boolean;
+  isFile: boolean;
+}
+
+/**
+ * Information about a file in the in-memory filesystem.
+ */
+interface FsStat {
+  exists: boolean;
+  isFile: boolean;
+  isDirectory: boolean;
+  size: number;
+}
+
+/**
+ * Create an in-memory FS which proxies the underlying OS filesystem using an
+ * in-memory cache. FS writes can accumulate on the in-memory system, using an
+ * API similar to Node.js' `"fs"` module, and then be committed to disk as a
+ * unit.
+ *
+ * Files written to the in-memory system can be edited, deleted, and so on.
+ * This allows the compiler to proceed freely as if it is modifying the
+ * filesystem, modifying the world in whatever way suits it, while deferring
+ * actual FS writes until the end of the compilation process, making actual
+ * changes to the filesystem on disk contingent on an error-free build or any
+ * other condition.
+ *
+ * @param sys a compiler system object
+ * @returns an in-memory filesystem interface
+ */
 export const createInMemoryFs = (sys: d.CompilerSystem) => {
-  const items: d.FsItems = new Map();
+  /**
+   * Map to hold the items in the in-memory cache which proxies the underlying
+   * OS filesystem.
+   */
+  const items: FsItems = new Map();
   const outputTargetTypes = new Map<string, string>();
 
-  const accessData = async (filePath: string) => {
+  /**
+   * Check if a file exists at a provided path. This function will attempt to
+   * use the in-memory cache before performing a blocking read. In the event of
+   * a cache hit, the content from the cache will be returned and the read skipped.
+   *
+   * @param filePath the path to the file to read
+   * @returns `true` if the file exists, `false` otherwise
+   */
+  const access = async (filePath: string): Promise<boolean> => {
     const item = getItem(filePath);
-    if (typeof item.exists === 'boolean') {
-      return {
-        exists: item.exists,
-        isDirectory: item.isDirectory,
-        isFile: item.isFile,
-      };
-    }
-
-    const data = {
-      exists: false,
-      isDirectory: false,
-      isFile: false,
-    };
 
-    const s = await stat(filePath);
-    if (s) {
-      item.exists = s.exists;
-      item.isDirectory = s.isDirectory;
-      item.isFile = s.isFile;
-
-      data.exists = item.exists;
-      data.isDirectory = item.isDirectory;
-      data.isFile = item.isFile;
-    } else {
-      item.exists = false;
```

Comment:
actually scratch that — it's been a while so I forgot, but the reason that I removed this code (`accessData`) was because it calls the `stat` function and, afaict, the assignments which modify the `item` returned by `getItem` are redundant with what is done there, see:

https://github.com/ionic-team/stencil/blob/2e773244f43cee77febb1715d1393bfa4df1309f/src/compiler/sys/in-memory-fs.ts#L558-L589

there we already have a series of checks which ensure that `.isFile`, `.isDirectory`, and so on are properly set

Reply:
Ah! Good catch!